### PR TITLE
Benchmarking only iree_vm_invoke moving everything else outside registred function

### PR DESCRIPTION
**Pull request by asaadaldien**
Benchmarks invokes the registered functions multiple times (not only the benchmarking loop) resulting creating IREE driver, module...etc over and over. It also makes it impossible to use something like `simpleperf` with current LLVMAOT because traces will be scattered from different dylib executables.

### CODE REVIEW

**Comment by GMNGeoffrey** [comment]
Only the input creation and iree invoke is in the inner benchmark loop https://github.com/google/iree/blob/main/iree/tools/iree-benchmark-module-main.cc#L141=L151. The benchmark framework does invoke the benchmark function a couple times though (I think 3 or so). Is that what you're referring to?

**Comment by GMNGeoffrey** [comment]
Oh missed that this was a PR, not an issue...

File: iree/tools/iree-benchmark-module-main.cc
Comment (GMNGeoffrey) [inline iree/tools/iree-benchmark-module-main.cc:146]: Could you move this back into a separate function? This is pretty noise for main IMO and also defeats status usage

Comment (asaadaldien) [reply iree/tools/iree-benchmark-module-main.cc:146]: By this do you mean everything (creating device, module...etc) ?  `IREE_CHECK_OK(...);` asserts prints which like so its actually clear what is going on when it fails.

Comment (GMNGeoffrey) [reply iree/tools/iree-benchmark-module-main.cc:146]: Yes. Having it in a separate function was deliberate to allow for status usage (and avoid a giant main function). Now you've got manually returning exit codes

File: iree/tools/iree-benchmark-module-main.cc
Comment (GMNGeoffrey) [inline iree/tools/iree-benchmark-module-main.cc:223]: This is super weird... You're passing the status to another function so that that function can do a check on it?

Comment (asaadaldien) [reply iree/tools/iree-benchmark-module-main.cc:223]: If we checked the status here we will fail early and something like `list-benchmarks` won't work

Comment (GMNGeoffrey) [reply iree/tools/iree-benchmark-module-main.cc:223]: Ah ok. I think that's worth a comment.

Also why do these have to go from StatusOr to Status with an output arg? Can we not thread through the StatusOr in the same way?

**Review by GMNGeoffrey — APPROVED** [review]
Sorry for the delayed review here. Approving to unblock you. I still think the factoring here is a bit strange, but I'll stop bike shedding.

**Comment by hanhanW** [comment]
Just found that this wasn't landed before https://github.com/google/iree/pull/3330

The flag names are changed.

input_file -> module_file
inputs -> function_inputs
inputs_file -> function_inputs_file

**Comment by GMNGeoffrey** [comment]
It looks like the GPU build broke with this. https://source.cloud.google.com/results/invocations/bfcdc8b4-4914-4720-99e3-07251345fd73/targets/iree%2Fgcp_ubuntu%2Fcmake%2Flinux%2Fx86-turing%2Fmain/log Looks like it may actually be related because it's specifically in the usage of iree-benchmark-module

**Comment by hanhanW** [comment]
Didn't look into detail. I suspect that this is because the order of releasing resource. I'm confused that why we didn't catch this before merging the PR.

**Comment by asaadaldien** [comment]
Strange! I can run the failing test on main with swiftshader fine on my local machine!
```
 bazel run iree/tools/test:simple.mlir.test
```

**Comment by GMNGeoffrey** [comment]
Yeah swiftshader is passing. It's only on the real gpu that it's failing

**Comment by asaadaldien** [comment]
I can repro now with physical gpu device. @hanhanW  The change is releasing the resources is the same order as before . I don't see anything obvious/changed other than multiple `iree_vm_invoke` with same context. If that is the problem is there a way to block until all `iree_vm_invoke` are done?

**Comment by hanhanW** [comment]
Looking. Seems that it's passing if run the command manually.

I also remove the cleanup and run `iree-benchmark-module` on the actual module file and it passed.

**Comment by hanhanW** [comment]
oh it didn't pass. It somehow crash in the end.

```
Running build/iree/tools/iree-benchmark-module
Run on (72 X 3700 MHz CPU s)
CPU Caches:
  L1 Data 32 KiB (x36)
  L1 Instruction 32 KiB (x36)
  L2 Unified 1024 KiB (x36)
  L3 Unified 25344 KiB (x2)
Load Average: 0.43, 0.64, 1.40
***WARNING*** CPU scaling is enabled, the benchmark real time measurements may be noisy and will incur extra overhead.
------------------------------------------------------------------------
Benchmark                              Time             CPU   Iterations
------------------------------------------------------------------------
BM_abs/process_time/real_time      0.215 ms        0.178 ms         3940
zsh: segmentation fault  build/iree/tools/iree-benchmark-module --driver=vulkan --entry_function=abs
```

**Comment by hanhanW** [comment]
Verified that it is able to run with `iree-run-module`.

It even fails when comment out `iree::RegisterModuleBenchmarks` and `::benchmark::RunSpecifiedBenchmarks`.

**Comment by hanhanW** [comment]
> Verified that it is able to run with `iree-run-module`.
> 
> It even fails when comment out `iree::RegisterModuleBenchmarks` and `::benchmark::RunSpecifiedBenchmarks`.

The stack trace (from gdb) is:

```
#0  __GI___pthread_mutex_lock (mutex=0x58) at ../nptl/pthread_mutex_lock.c:67
#1  0x00007ffff25c1003 in ?? () from /lib/x86_64-linux-gnu/libEGL_mesa.so.0
#2  0x00007ffff4846c1b in eglReleaseThread () from /lib/x86_64-linux-gnu/libEGL.so.1
#3  0x00007ffff69c5db7 in ?? () from /lib/x86_64-linux-gnu/libGLX_nvidia.so.0
#4  0x00007ffff69c4bc9 in ?? () from /lib/x86_64-linux-gnu/libGLX_nvidia.so.0
#5  0x00007ffff7fe2365 in _dl_fini () at dl-fini.c:143
#6  0x00007ffff7a785a7 in __run_exit_handlers (status=0x0, listp=0x7ffff7bf8718 <__exit_funcs>, run_list_atexit=run_list_atexit@entry=0x1, run_dtors=run_dtors@entry=0x1) at exit.c:108
#7  0x00007ffff7a7874a in __GI_exit (status=<optimized out>) at exit.c:139
#8  0x00007ffff7a60cd1 in __libc_start_main (main=0x45c960 <main(int, char**)>, argc=0x5, argv=0x7fffffffddc8, init=<optimized out>, fini=<optimized out>, rtld_fini=<optimized out>, stack_end=0x7fffffffddb8) at ../csu/libc-start.c:342
```

which looks like https://github.com/google/iree/issues/2659

**Comment by hanhanW** [comment]
Fixing this in https://github.com/google/iree/pull/3377

**Pull request merged**
